Adds struct variant to `regexp_extract_all`

J-Meyers opened the pull request:
Duckdb has a really neat variant of `regexp_extract` which allows providing column names then is able to export each of the capture groups after a single regex match, this really neat feature wasn't supported when extracting all results. 

There isn't an easy way to do this otherwise, especially without at least one more call to match.

This adds that functionality and tries to share some code for parameter parsing with `regexp_extract`. 

I tried to add a variety of tests, but with stuff like this there are quite a few cases so I'm very open to adding any more.

I encountered this while trying to do some processing from a generic `read_text` function, I also ran into wanting to have a `regexp_instr` function like snowflake and others.

It is possible to add support for a variable pattern, but it makes the logic a bit more complicated, there are a few options for how to  handle differing number of capture groups, not supporting it I believe is consistent with the `regexp_extract` function.

J-Meyers committed:
Add regex_extract_all (struct variant)

J-Meyers committed:
Correct bugs with parameter checking

J-Meyers committed:
Formatting

J-Meyers committed:
Clang-tidy

lnkuiper commented on src/function/scalar/string/regexp/regexp_util.cpp:
```diff
@@ -78,6 +80,50 @@ void ParseRegexOptions(ClientContext &context, Expression &expr, RE2::Options &t
+		}
+		if ((idx_t)group_cnt < out_names.size()) {
+			throw BinderException("Not enough capturing groups (%d) for provided names (%llu)", group_cnt,
+			                      (uint64_t)out_names.size());
```
Could you please avoid C-style casts and use `NumericCast` instead?

lnkuiper reviewed (changes requested):
Thanks for the PR! This looks excellent. I just have a few minor nitpicks, then I think this should be ready to go:

J-Meyers committed:
PR feedback (test exceptions, casting, function move)

J-Meyers committed:
Merge remote-tracking branch 'origin/main' into regexp_functions

J-Meyers commented:
@lnkuiper I believe I've addressed your comments, however both times the CI tried to run the job `LinuxRelease / Linux CLI (arm64) (pull_request)` Was not picked up and is marked as failing and gave error: `The job was not acquired by Runner of type hosted even after multiple attempts` I merged in main to try and trigger it again, but faced the same issue

lnkuiper reviewed (approved):
Thanks for the changes! I've triggered a rerun of the CI job that GH actions wasn't able to do earlier. LGTM once that passes
